src/helpers/buffers.rs

```diff
@@ -3,13 +3,58 @@ use crate::protocol::RecordId;
 use std::collections::hash_map::Entry;
 use std::collections::HashMap;
 use std::mem;
+use std::mem::ManuallyDrop;
+
+use std::ops::Range;
+use thiserror::Error;
 use tokio::sync::oneshot;
 
+/// Accumulator of records that must be send down to network.
+#[derive(Debug)]
+struct SendBatch {
+    /// The current offset for records that can be accepted by this buffer.
+    /// If offset is 0, record identifiers from range 0..`batch_size` are accepted.
+    /// If offset is 1, record identifiers from `batch_size`..2*`batch_size` are accepted
+    offset: u32,
+
+    /// number of records to accumulate inside the buffer (per channel) before flush.
+    batch_size: u32,
+
+    /// The range of valid identifiers currently accepted by the batch.
+    accepted_range: Range<RecordId>,
+
+    /// Tracks number of elements in `data`. Must be stored separately because `data` always
+    /// contains elements
+    len: u32,
+
+    /// Vector of messages with holes. At initialization time it contains `batch_size`
+    /// elements, every element is set to `None`. As vector fills in, elements are replaced with
+    /// actual messages. Each message takes position in this vector according to its `record_id`
+    /// value.
+    ///
+    /// Here is an example layout for a half-filled batch with `batch_size` = 5 and `offset` = 1.
+    /// Note that only record identifiers are shown because other information is irrelevant for this
+    /// example.
+    /// [None, Record(6), None, Record(8), Record(9)], len = 3
+    ///
+    /// Once every element in the vec is set, it is considered full and data can be moved
+    /// away from it. `SendBatch` guarantees no extra allocations in this case because this vector
+    /// is just reinterpreted as `Vec<MessageEnvelope>`.
+    ///
+    /// If `data` is not full and buffer needs to be flushed to network, a new vector will be
+    /// allocated and `data` will be copied element-by-element which is less efficient and therefore
+    /// should be avoided if possible
+    ///
+    /// This implementation guarantees that this vector is sorted according to `MessageEnvelope::record_id`
+    data: Vec<Option<MessageEnvelope>>,
```

Comment:
I think that this could also use a type of `[Option<Box<[u8]>>; N]` where `N` is twice the batch size.  If everything is working properly, you will want to be able to receive messages that are out of order by some number.  Ideally, that is a small number, at least smaller than the batch size.

But you want to be able to eject a slice from this store for sending when it is ready, while you are also receiving new messages to send in the next batch.  That suggests that you need a store that is large enough for a whole batch, plus whatever your reordering tolerance is.

But you also want this store to be a clean multiple of the batch size so that you don't have to worry about weird offsets into the store.  (Maybe you can do that for that exceptional case you talk about in the comment, but it's OK if that costs a lot.)

An alternative is to maintain two buffers, b1 and b2, with each maintaining an initial offset (or with a single starting offset and a flag saying which one is primary, or you can mem::swap pointers to them, I don't care).

Reply:
Another way of approaching this is to change the message so that each item is fixed length.  Then, you can store a byte array of fixed size, with an associated bit vector (BitVec from the bit_vec crate perhaps?) that indicates which values are yet to be populated.  Then, writing to the network only involves handing off a reference to the slice of bytes.

I might prefer that approach best in fact.  It's much simpler than your current design.  It also doesn't really pay a huge tax for the case where you need to write out some of the values.

Reply:
Note that you can move the buffer size into the type definition using const generics.  I don't know if that is worthwhile, but you might consider it if moving to a message type that is fixed size.

Reply:
I like the fixed length approach. Do you have a suggestion for how to pick the appropriate message length? we seem to send messages that carry either 1 or 2 field values, but I was assuming that we may send a whole batch of values (permutation)? but even to decide between 1 and 2, we would need to either allocate space for the largest possible message (50% overhead for 1 field value messages) or somehow deliver this metadata to the buffer definition

We could use const generic parameter for `Step` but generic  expressions are not a thing yet so we can't simply say `payload: [u8; Step::MESSAGE_SIZE]`...

Reply:
Yeah, I don't think that we need to have the buffer be declared as `[u8; N]`; we can probably just use `Vec<u8>` and allocate it at runtime, right at the start.  For now, we can just have the message type implement a trait that has a const for its encoded length (in bits eventually perhaps, but start with bytes) and has an `write_to(&self, &mut [u8])` or `write_to(&self, buf: &mut [u8], offset: usize)` method or something like that.

Reply:
Fixed size messages appears to be working (I picked 8 bytes for message size for now and use in-place buffers) in my prototype. I also realized that writing out just some values from the buffer wouldn't work without extra bookkeeping: we can't really move the window until *all* messages have been received. It may be possible to keep the bit vec around until all records have been received, but I don't see a clear benefit for it as buffer capacity is not increased anyway. 

Wondering if I should just remove this logic to write out partial buffers and try to mitigate the issues with `Attribution` by keeping buffers that are multiples of `batch_size` and multiplier is > 1